---
title: "Apparition of a leaky y in a population"
author: "Frederic Michaud"
date: "`r Sys.Date()`"
output:
  rmarkdown::pdf_document: default
  rmarkdown::html_vignette: default
vignette: >
  %\VignetteIndexEntry{Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

In this notebook, we consider a x-y system with a leaky-y appearing at some point. We will briefly explore under what circonstances the leaky-y invade the system and in which case it will not. What we mean by a leaky-y is the following: if an individual receive a leaky-y, it becomes usualy a male, but sometimes (here 10% of the time) it can turn into a female. 

In the absence of any selection, a leaky-y is neutral. If it appears, its evolution only rely on drift. Therefore, in an infinite population, we expect it to remain constant. Under a selection which is unrealted to the sex, we expect the same to happen. One scenario where this could be different is if the sexual chromosom have accumulated sex-antagonisit gene. In this case, females with male benefitial appear and the selection on the leaky-y is different than on the y. 
     
     
##First case   
In the first case, we consider a symetric SA gene   
     
     
### setting the genome   
To specify a leaky-y, all we need to do is add a third allele (L like Leaky-y) to the x-y locus, and specify for each combination of x-y-L, the proportion of individual turning into male and female. Here, we expect xx individual to be always female, xy and yL to be always male and xL to be female 10% of the time and male 90% of the time. This is what is written in the vector sd. 1 means purely male, 0 purely female and any value inbetween the frequency of individual turning into male. Notice that here, no yy appears so we do not need to specify it. On the other hand, xL female will mate with xy male, generating yL individual. If we forget to specify such individuals, the program act weirdly, and result should not be trusted.   

We also add a second locus with a sex-antagonistic gene (SA gene).
```{r}
locus1 = create.locus(allele1=c(1,1,1,2,3),
                      allele2 = c(1,2,3,3,3),
                      sd = c(0,1,0.9,1,1),
                      allele.name = c("x","y","L"))
locus2 = create.locus(allele1=  c(1,1,2),
                      allele2 = c(1,2,2),
                      fitness.female = c(1,0.9,0.8),
                      fitness.male = c(0.8,0.9,1),
                      allele.name = c("F","M"))
genome = create.genome(locus = list(locus1,locus2))
genome
```

### Simulating the evolution and plotting result   
Since we want to simulate the invasion of a Leaky-y, it should appears with a very low frequency. This can be done using the function `get.frequency.from.one.allele.frequency` which allows to get initial frequency with one allele being very rare. 
```{r}
 #the third allele of the first locus appears with a frequency of 0.01
initial.frequency <- get.frequency.from.one.allele.frequency(genome,
                                                            locus = 1,
                                                             allele = 3,
                                                             allele.frequency =  0.01)
freqs <- compute.frequency.evolution(genome, initial.frequency, generations=1000)
plot.haplotype.frequency(genome,freqs)
```
    
So in the case of symmetric Sex-antagonisitc selection, a leaky-y does not invade the system. This make sens. For a symetric SA gene, X fix the benefitial for male, and y the benefitial for female. A leaky y would be counter-selected for because it would appears from time to time in female where it is deleterious, and would not bring advantage to the male carrying it As we will see, this argument is somehow correct but also a bit naive, since for some parameters it does not hold. 
   
### non-symetric SA   
We will now turn into a non-symetric SA gene. We know that in the case of the x-y system, this actually makes a difference in term of the dynamic. Will it be the same here? We will consider the same system as above, but just change the fitness of the SA
```{r}
locus2 = create.locus(allele1=  c(1,1,2),
                      allele2 = c(1,2,2),
                      fitness.female = c(1,0.9,0.8),
                      fitness.male = c(0.6,0.8,1),
                      allele.name = c("F","M"))
genome = create.genome(locus = list(locus1,locus2))
```
```{r}

initial.frequency <- get.frequency.from.one.allele.frequency(genome,locus = 1,
                                                             allele = 3,
                                                             allele.frequency =  0.01)
freqs <- compute.frequency.evolution(genome, initial.frequency, generations=10000)
plot.haplotype.frequency(genome,freqs)
```
    
We see that in this case, the leaky-y actually invade the system. This is a very slow process ( it takes around 10 generations for the ym to disapear, and 10'000 generation for the leaky-y to reach its equilibrium value, so a factor 1000 in speed). So what is the difference between this case and the one we considered previously? One big difference is that here we have a polymorphism maintain on the X chromosome. So one difference is that a leaky-y is less costly to female than it was before. But we still don't know why a leaky-y should be selected. One reason is probably that leaky-y allow to make yL and LL male, who are homozygothe for the male beneficial allele. This male we be strongly selected, and probably allows the leaky-y to invade the population. 
